File: CropawayWindows/CropawayWindows/ViewModels/CropEditorViewModel.cs
using System.Windows;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using CropawayWindows.Models;

namespace CropawayWindows.ViewModels;

public partial class CropEditorViewModel : ObservableObject
{
    [ObservableProperty]
    private CropMode _mode = CropMode.Rectangle;

    [ObservableProperty]
    private bool _isEditing = true;

    // Rectangle crop (normalized 0-1)
    [ObservableProperty]
    private Rect _cropRect = new(0, 0, 1, 1);

    // Edge crop
    [ObservableProperty]
    private Models.EdgeInsets _edgeInsets = new();

    // Circle crop
    [ObservableProperty]
    private Point _circleCenter = new(0.5, 0.5);

    [ObservableProperty]
    private double _circleRadius = 0.4;

    // Freehand
    [ObservableProperty]
    private List<Point> _freehandPoints = new();

    [ObservableProperty]
    private byte[]? _freehandPathData;

    [ObservableProperty]
    private bool _isDrawing;

    // AI mask
    [ObservableProperty]
    private byte[]? _aiMaskData;

    [ObservableProperty]
    private List<AIPromptPoint> _aiPromptPoints = new();

    [ObservableProperty]
    private string? _aiTextPrompt;

    [ObservableProperty]
    private Rect _aiBoundingBox;

    [ObservableProperty]
    private AIInteractionMode _aiInteractionMode = AIInteractionMode.Point;

    // Export settings per video
    [ObservableProperty]
    private bool _preserveWidth = true;

    [ObservableProperty]
    private bool _enableAlphaChannel;

    // Active video
    private VideoItem? _currentVideo;
    private bool _isSyncing;

    // Callback for when crop editing ends (drag gesture completed)
    public event Action? CropEditEnded;

    public void BindTo(VideoItem video)
    {
        _currentVideo = video;
        var config = video.CropConfig;

        _isSyncing = true;
        Mode = config.Mode;
        CropRect = config.CropRect;
        EdgeInsets = config.EdgeInsets;
        CircleCenter = config.CircleCenter;
        CircleRadius = config.CircleRadius;
        FreehandPoints = config.FreehandPoints.ToList();
        FreehandPathData = config.FreehandPathData;
        AiMaskData = config.AiMaskData;
        AiPromptPoints = config.AiPromptPoints.ToList();
        AiTextPrompt = config.AiTextPrompt;
        AiBoundingBox = config.AiBoundingBox;
        AiInteractionMode = config.AiInteractionMode;
        PreserveWidth = config.PreserveWidth;
        EnableAlphaChannel = config.EnableAlphaChannel;
        _isSyncing = false;
    }

    // Sync changes back to CropConfiguration
    partial void OnModeChanged(CropMode value)
    {
        if (!_isSyncing && _currentVideo != null)
            _currentVideo.CropConfig.Mode = value;
    }

    partial void OnCropRectChanged(Rect value)
    {
        if (!_isSyncing && _currentVideo != null)
            _currentVideo.CropConfig.CropRect = value;
    }

    partial void OnCircleCenterChanged(Point value)
    {
        if (!_isSyncing && _currentVideo != null)
            _currentVideo.CropConfig.CircleCenter = value;
    }

    partial void OnCircleRadiusChanged(double value)
    {
        if (!_isSyncing && _currentVideo != null)
            _currentVideo.CropConfig.CircleRadius = value;
    }

    partial void OnFreehandPointsChanged(List<Point> value)
    {
        if (!_isSyncing && _currentVideo != null)
            _currentVideo.CropConfig.FreehandPoints = value;
    }

    partial void OnFreehandPathDataChanged(byte[]? value)
    {
        if (!_isSyncing && _currentVideo != null)
            _currentVideo.CropConfig.FreehandPathData = value;
    }

    partial void OnAiMaskDataChanged(byte[]? value)
    {
        if (!_isSyncing && _currentVideo != null)
            _currentVideo.CropConfig.AiMaskData = value;
    }

    partial void OnAiBoundingBoxChanged(Rect value)
    {
        if (!_isSyncing && _currentVideo != null)
            _currentVideo.CropConfig.AiBoundingBox = value;
    }

    partial void OnPreserveWidthChanged(bool value)
    {
        if (!_isSyncing && _currentVideo != null)
            _currentVideo.CropConfig.PreserveWidth = value;
    }

    partial void OnEnableAlphaChannelChanged(bool value)
    {
        if (!_isSyncing && _currentVideo != null)
            _currentVideo.CropConfig.EnableAlphaChannel = value;
    }

    // Get effective crop area for current mode
    public Rect EffectiveCropRect
    {
        get
        {
            return Mode switch
            {
                CropMode.Rectangle => CropRect,
                CropMode.Circle => new Rect(
                    CircleCenter.X - CircleRadius,
                    CircleCenter.Y - CircleRadius,
                    CircleRadius * 2,
                    CircleRadius * 2),
                CropMode.Freehand when FreehandPoints.Count > 0 => GetFreehandBounds(),
                CropMode.AI when AiBoundingBox.Width > 0 => AiBoundingBox,
                _ => new Rect(0, 0, 1, 1)
            };
        }
    }

    private Rect GetFreehandBounds()
    {
        if (FreehandPoints.Count == 0) return new Rect(0, 0, 1, 1);
        var minX = FreehandPoints.Min(p => p.X);
        var maxX = FreehandPoints.Max(p => p.X);
        var minY = FreehandPoints.Min(p => p.Y);
        var maxY = FreehandPoints.Max(p => p.Y);
        return new Rect(minX, minY, maxX - minX, maxY - minY);
    }

    [RelayCommand]
    public void Reset()
    {
        CropRect = new Rect(0, 0, 1, 1);
        EdgeInsets = new Models.EdgeInsets();
        CircleCenter = new Point(0.5, 0.5);
        CircleRadius = 0.4;
        FreehandPoints = new List<Point>();
        FreehandPathData = null;
        AiMaskData = null;
        AiPromptPoints = new List<AIPromptPoint>();
        AiTextPrompt = null;
        AiBoundingBox = default;

        if (_currentVideo != null)
        {
            _currentVideo.CropConfig.Reset();
        }
    }

    // Freehand drawing
    public void StartDrawing(Point point)
    {
        FreehandPoints = new List<Point> { point };
        IsDrawing = true;
    }

    public void ContinueDrawing(Point point)
    {
        if (!IsDrawing) return;
        var pts = FreehandPoints.ToList();
        pts.Add(point);
        FreehandPoints = pts;
    }

    public void EndDrawing()
    {
        IsDrawing = false;
        if (FreehandPoints.Count > 2)
        {
            var first = FreehandPoints[0];
            var last = FreehandPoints[^1];
            var distance = Math.Sqrt(Math.Pow(first.X - last.X, 2) + Math.Pow(first.Y - last.Y, 2));
            if (distance > 0.01)
            {
                var pts = FreehandPoints.ToList();
                pts.Add(first);
                FreehandPoints = pts;
            }
        }
    }

    [RelayCommand]
    public void ClearFreehand()
    {
        FreehandPoints = new List<Point>();
        FreehandPathData = null;
    }

    [RelayCommand]
    public void ClearAIMask()
    {
        AiMaskData = null;
        AiPromptPoints = new List<AIPromptPoint>();
        AiTextPrompt = null;
        AiBoundingBox = default;
    }

    public void NotifyCropEditEnded()
    {
        CropEditEnded?.Invoke();
    }

    [RelayCommand]
    private void SetRectangleMode() => Mode = CropMode.Rectangle;

    [RelayCommand]
    private void SetCircleMode() => Mode = CropMode.Circle;

    [RelayCommand]
    private void SetFreehandMode() => Mode = CropMode.Freehand;

    [RelayCommand]
    private void SetAIMode() => Mode = CropMode.AI;
}

File: cropaway/ViewModels/CropEditorViewModel.swift
//
//  CropEditorViewModel.swift
//  cropaway
//

import Combine
import Foundation
import SwiftUI
import CoreGraphics

@MainActor
final class CropEditorViewModel: ObservableObject {
    @Published var mode: CropMode = .rectangle
    @Published var isEditing: Bool = true
    @Published var isDragging: Bool = false  // True while user is actively dragging a handle

    // Rectangle crop (normalized 0-1)
    @Published var cropRect: CGRect = CGRect(x: 0, y: 0, width: 1, height: 1)

    // Edge crop
    @Published var edgeInsets: EdgeInsets = EdgeInsets()

    // Circle crop
    @Published var circleCenter: CGPoint = CGPoint(x: 0.5, y: 0.5)
    @Published var circleRadius: Double = 0.4

    // Freehand
    @Published var freehandPoints: [CGPoint] = []
    @Published var freehandPathData: Data? = nil  // Full bezier vertex data
    @Published var isDrawing: Bool = false

    // AI mask (fal.ai)
    @Published var aiMaskData: Data?
    @Published var aiPromptPoints: [AIPromptPoint] = []
    @Published var aiTextPrompt: String?
    @Published var aiBoundingBox: CGRect = .zero
    @Published var aiInteractionMode: AIInteractionMode = .point

    // Callback for when crop editing ends (drag gesture completed)
    // Used for auto-keyframe creation
    var onCropEditEnded: (() -> Void)?

    // Active video
    private var currentVideo: VideoItem?
    private var cancellables = Set<AnyCancellable>()

    func bind(to video: VideoItem) {
        cancellables.removeAll()
        currentVideo = video

        let config = video.cropConfiguration

        // Sync from config
        mode = config.mode
        cropRect = config.cropRect
        edgeInsets = config.edgeInsets
        circleCenter = config.circleCenter
        circleRadius = config.circleRadius
        freehandPoints = config.freehandPoints
        freehandPathData = config.freehandPathData
        aiMaskData = config.aiMaskData
        aiPromptPoints = config.aiPromptPoints
        aiTextPrompt = config.aiTextPrompt
        aiBoundingBox = config.aiBoundingBox
        aiInteractionMode = config.aiInteractionMode

        // Sync changes back to config
        $mode
            .dropFirst()
            .sink { config.mode = $0 }
            .store(in: &cancellables)

        $cropRect
            .dropFirst()
            .sink { config.cropRect = $0 }
            .store(in: &cancellables)

        $edgeInsets
            .dropFirst()
            .sink { config.edgeInsets = $0 }
            .store(in: &cancellables)

        $circleCenter
            .dropFirst()
            .sink { config.circleCenter = $0 }
            .store(in: &cancellables)

        $circleRadius
            .dropFirst()
            .sink { config.circleRadius = $0 }
            .store(in: &cancellables)

        $freehandPoints
            .dropFirst()
            .sink { config.freehandPoints = $0 }
            .store(in: &cancellables)

        $freehandPathData
            .dropFirst()
            .sink { config.freehandPathData = $0 }
            .store(in: &cancellables)

        $aiMaskData
            .dropFirst()
            .sink { config.aiMaskData = $0 }
            .store(in: &cancellables)

        $aiPromptPoints
            .dropFirst()
            .sink { config.aiPromptPoints = $0 }
            .store(in: &cancellables)

        $aiTextPrompt
            .dropFirst()
            .sink { config.aiTextPrompt = $0 }
            .store(in: &cancellables)

        $aiBoundingBox
            .dropFirst()
            .sink { config.aiBoundingBox = $0 }
            .store(in: &cancellables)

        $aiInteractionMode
            .dropFirst()
            .sink { config.aiInteractionMode = $0 }
            .store(in: &cancellables)
    }

    // Get effective crop area for current mode
    var effectiveCropRect: CGRect {
        switch mode {
        case .rectangle:
            return cropRect
        case .circle:
            let diameter = circleRadius * 2
            return CGRect(
                x: circleCenter.x - circleRadius,
                y: circleCenter.y - circleRadius,
                width: diameter,
                height: diameter
            )
        case .freehand:
            guard !freehandPoints.isEmpty else { return cropRect }
            let xs = freehandPoints.map { $0.x }
            let ys = freehandPoints.map { $0.y }
            let minX = xs.min() ?? 0
            let maxX = xs.max() ?? 1
            let minY = ys.min() ?? 0
            let maxY = ys.max() ?? 1
            return CGRect(x: minX, y: minY, width: maxX - minX, height: maxY - minY)
        case .ai:
            return aiBoundingBox.width > 0 ? aiBoundingBox : CGRect(x: 0, y: 0, width: 1, height: 1)
        }
    }

    func reset() {
        cropRect = CGRect(x: 0, y: 0, width: 1, height: 1)
        edgeInsets = EdgeInsets()
        circleCenter = CGPoint(x: 0.5, y: 0.5)
        circleRadius = 0.4
        freehandPoints = []
        freehandPathData = nil
        aiMaskData = nil
        aiPromptPoints = []
        aiTextPrompt = nil
        aiBoundingBox = .zero
    }

    // Freehand drawing
    func startDrawing(at point: CGPoint) {
        freehandPoints = [point]
        isDrawing = true
    }

    func continueDrawing(to point: CGPoint) {
        guard isDrawing else { return }
        freehandPoints.append(point)
    }

    func endDrawing() {
        isDrawing = false
        // Close the path by adding start point if needed
        if let first = freehandPoints.first, let last = freehandPoints.last {
            if first.distance(to: last) > 0.01 {
                freehandPoints.append(first)
            }
        }
    }

    func clearFreehand() {
        freehandPoints = []
        freehandPathData = nil
    }

    /// Called when a crop editing gesture ends (drag completed)
    /// This triggers the onCropEditEnded callback for auto-keyframe creation
    func notifyCropEditEnded() {
        onCropEditEnded?()
    }

    // MARK: - AI Segmentation

    /// Clear AI mask data
    func clearAIMask() {
        aiMaskData = nil
        aiPromptPoints = []
        aiTextPrompt = nil
        aiBoundingBox = .zero
    }
}
